test_scripts/collect_data.py

```diff
@@ -0,0 +1,52 @@
+import os
+from time import sleep
+import subprocess
+import re
+
+run_time = 10
+reset_time = 5
+pps_sample = [10000,50000,100000]
+pcycle_sample = [20,200,2000]
+num_runs = 3
+
+
+def get_hw_latency(text):
```

Comment:
One possible bug is that the 'get_summary' may cause race conditions when it reads the circular buffer.
Is there a way to make sure that the worker core won't write to the buffer when we call 'get_summary'.

Reply:
As per the documentation of boost::circular_buffer the iterator will return valid values as long as the iterator points to a valid location, thus in case the read races with the write it will read the latest value which is fine in this case.

Reply:
"If multiple threads access a single circular_buffer, and at least one of the threads may potentially write, then the user is responsible for ensuring mutual exclusion between the threads during the container accesses."
In our case, the CommandGetSummary may run on a different gRPC thread, while the worker threads writes to the buffer. Then, race conditions can happen, which leads to undefined bahaviors. I am okay with this non-lock implementation as 'get_summary' will not get called during experiments. However, in your test, I think you'd use a lock to protect the buffer. Otherwise, the buffer data may be corrupted.

Source: https://www.boost.org/doc/libs/1_62_0/doc/html/circular_buffer/implementation.html